Migrate addon a11y to TS

gaetanmaisse opened the pull request:
Issue: https://github.com/storybooks/storybook/issues/5030

## What I did

 - Migrate `@storybook/addon-a11y` to TypeScript
 - Refactor some imports/exports
 - Add some types and interfaces and use them to type codebase

gaetanmaisse committed:
chore: add tsconfig in addons/a11y

gaetanmaisse committed:
chore: migrate src of addons/a11y to Typescript

gaetanmaisse committed:
chore: type src of addons/a11y

gaetanmaisse committed:
chore: remove prop-types dependency from package.json of addons/a11y

gaetanmaisse committed:
chore: use named import/export instead of default ones in addons/a11y

Keraito commented on addons/a11y/src/components/ColorBlindness.tsx:
```diff
@@ -83,7 +93,7 @@ class ColorBlindness extends Component {
         trigger="click"
         tooltipShown={expanded}
-        onVisibilityChange={s => this.setState({ expanded: s })}
+        onVisibilityChange={(s: boolean) => this.setState({ expanded: s })}
```
To prevent [performance issue](https://reactjs.org/docs/faq-functions.html#arrow-function-in-render) down the road, please put it into a [class property](https://reactjs.org/docs/faq-functions.html#class-properties-stage-3-proposal).

Keraito commented on addons/a11y/src/components/Tabs.tsx:
```diff
@@ -35,7 +34,7 @@ const Item = styled.button(
     },
   }),
-  ({ active, theme }) =>
+  ({ active, theme }: any) =>
```
we can at least say active is boolean right?

gaetanmaisse replied:
Unfortunately same as https://github.com/storybooks/storybook/pull/5738#discussion_r260462864 :/ 

So I will take a look to `@storybook/theming` really fast.

Keraito replied:
Even just `{active:boolean, theme:any}` would be stricter, does that also error?

gaetanmaisse replied:
Yes as we are destructuring an object with other properties, I also tried `{active:boolean; theme:any; [key: string]: any}` but still get TS errors... I can not do better than `any` for now

Keraito reviewed (approved):
Looks good, a few small comments so please look into those! 😊

gaetanmaisse commented on addons/a11y/src/components/A11YPanel.tsx:
```diff
@@ -1,23 +1,22 @@
     marginRight: '4px',
   },
-  ({ status, theme }) =>
+  ({ status, theme }: any) =>
```
Yep! I was a bit lazy here because I think it should be automatically inferred by TS, I will investigate why it's not. I think we can do better typings in `@storybook/theming` ;)

gaetanmaisse replied:
Ok, it looks like I can not do better for now, if I try a stronger typings I got TS errors elsewhere...

gaetanmaisse commented on addons/a11y/src/components/A11YPanel.tsx:
```diff
@@ -33,17 +32,23 @@ const Violations = styled.span(({ theme }) => ({
 
-  state = {
+export class A11YPanel extends Component<A11YPanelProps, A11YPanelState> {
+  state: A11YPanelState = {
```
I had to because if I don't TS throws errors like: `TS7018: Object literal's property 'violations' implicitly has an 'any[]' type.`

gaetanmaisse committed:
refactor: use a class property to prevent performance issue

See https://reactjs.org/docs/faq-functions.html#arrow-function-in-render
And https://reactjs.org/docs/faq-functions.html#class-properties-stage-3-proposal

gaetanmaisse committed:
fix: fix  a11y option attribute in examples/official-storybook

shilman commented:
FYI I'm patching this back into `5.0.x` because the a11y bugfixes depend on this

shilman commented:
I take it back. a11y bugfixes will be available in 5.1 only
